src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -766,111 +760,184 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
+                        if (_lastPendingWriterShouldFlush)
+                        {
+                            await FlushOutgoingBytesAsync().ConfigureAwait(false);
+                        }
+
+                        // Wait for work to be added.
+                        await waitForSignal.ConfigureAwait(false);
+
+#if DEBUG
+                        lock (SyncObject)
+                        {
+                            Debug.Assert(_abortException is not null || _writeQueue.Count > 0, "Write queue signalled but no work to do ??");
+                        }
+#endif
+                    }
+                    else
+                    {
+                        if (writeEntry!.TryDisableCancellation())
+                        {
+                            int writeBytes = writeEntry.WriteBytes;
+
+                            // If the buffer has already grown to 32k, does not have room for the next request,
+                            // and is non-empty, flush the current contents to the wire.
+                            int totalBufferLength = _outgoingBuffer.Capacity;
+                            if (totalBufferLength >= UnflushedOutgoingBufferSize)
+                            {
+                                int activeBufferLength = _outgoingBuffer.ActiveLength;
+                                if (writeBytes >= totalBufferLength - activeBufferLength)
+                                {
+                                    await FlushOutgoingBytesAsync().ConfigureAwait(false);
+                                }
+                            }
+
+                            _outgoingBuffer.EnsureAvailableSpace(writeBytes);
+
+                            try
+                            {
+                                if (NetEventSource.IsEnabled) Trace($"{nameof(writeBytes)}={writeBytes}");
+
+                                // Invoke the callback with the supplied state and the target write buffer.
+                                bool flush = writeEntry.InvokeWriteAction(_outgoingBuffer.AvailableMemorySliced(writeBytes));
+
+                                writeEntry.SetResult();
+
+                                _outgoingBuffer.Commit(writeBytes);
+                                _lastPendingWriterShouldFlush |= flush;
+                            }
+                            catch (Exception e)
+                            {
+                                writeEntry.SetException(e);
+                            }
+                        }
+                    }
+                }
+
+                // Connection is aborting, so fail any outstanding writes.
+                lock (SyncObject)
                 {
-                    _inProgressWrite = FlushOutgoingBytesAsync();
+                    Debug.Assert(_abortException is not null);
+
+                    while (_writeQueue.TryDequeue(out WriteQueueEntry? writeEntry))
+                    {
+                        if (writeEntry!.TryDisableCancellation())
```

Comment:
This `!` shouldn't be necessary.  Are you seeing a warning if you remove it?

Reply:
You're right, it's not. I think I just copy/pasted from the same check above.

So the compiler understands the semantics of TryDequeue?

Reply:
`Queue<T>.TryDequeue` is declared as:
```C#
public bool TryDequeue([MaybeNullWhen(false)] out T result)
```
The compiler understand the combination of a) that `MaybeNullWhen(false)` attribute and b) you declared the type of the Queue to be `Queue<WriteQueueEntry>` rather than `Queue<WriteQueueEntry?>`.  So it can see that you're only inside the while block if TryDequeue returned true, in which case it's going to be setting the argument to a non-null WriteQueueEntry.